Expands `manual_memcpy` to lint ones with loop counters

rail-rain opened the pull request:
Closes #1670

This PR expands `manual_memcpy` to lint ones with loop counters as described in https://github.com/rust-lang/rust-clippy/issues/1670#issuecomment-293280204

Although the current code is working, I have a couple of questions and concerns.

~~Firstly, I manually implemented `Clone` for `Sugg` because `AssocOp` lacks `Clone`. As `AssocOp` only holds an enum, which is `Copy`, as a value, it seems `AssocOp` can be `Clone`; but, I was not sure where to ask it. Should I make a PR to `rustc`?~~ The [PR]( https://github.com/rust-lang/rust/pull/73629) was made.

Secondly, manual copying with loop counters are likely to trigger `needless_range_loop` and `explicit_counter_loop` along with `manual_memcpy`; in fact, I explicitly allowed them in the tests. Is there any way to disable these two lints when a code triggers `manual_memcpy`?

And, another thing I'd like to note is that `Sugg` adds unnecessary parentheses when expressions with parentheses passed to its `hir` function, as seen here:

```
error: it looks like you're manually copying between slices
  --> $DIR/manual_memcpy.rs:145:14
   |
LL |     for i in 3..(3 + src.len()) {
   |              ^^^^^^^^^^^^^^^^^^ help: try replacing the loop by: `dst[3..((3 + src.len()))].clone_from_slice(&src[..((3 + src.len()) - 3)])
```

However, using the `hir` function is needed to prevent the suggestion causing  errors when users use bitwise operations; and also this have already existed, for example: `verbose_bit_mask`. Thus, I think this is fine.

changelog: Expands `manual_memcpy` to lint ones with loop counters

rust-highfive commented:
r? @flip1995

(rust_highfive has picked a reviewer for you, use r? to override)

flip1995 commented:
cc https://github.com/rust-lang/rust/pull/73629

bors commented:
:umbrella: The latest upstream changes (presumably #5711) made this pull request unmergeable. Please resolve the merge conflicts.

rail-rain commented on clippy_lints/src/utils/sugg.rs:
```diff
@@ -263,21 +303,49 @@ impl<'a> Sugg<'a> {
 }
 
-impl<'a> std::ops::Not for Sugg<'a> {
+impl std::ops::Add<&Sugg<'_>> for Sugg<'_> {
```
I've tried it. Is this what you mean?

flip1995 commented on clippy_lints/src/utils/sugg.rs:
```diff
@@ -263,21 +304,53 @@ impl<'a> Sugg<'a> {
+    };
+}
+
+impl Add for &Sugg<'_> {
```
Yes that macro is what I meant. Can you also do this for the other operation?

rail-rain replied:
I've implemented `Neg` for `Sugg` without the reference versions, instead of direct calls of `make_binop` for now. Just to make it clear, do you mean overloading other operations like `Mul` for  future use regardless what this PR needs or implementing forwarding for unary operators? If the answer is the latter, as `make_unop` takes ownership, `self` have to be cloned to do so; or it will probably be possible to make `make_unop` taking references; which one do you think preferable?

flip1995 replied:
Ah I see, I should have looked closer at what is implemented for `Sugg`. In that case, you can keep it, as it is right now.

rail-rain replied:
I understand, I will do so. ...I mean making no change.

flip1995 commented:
@rail-rain `AssocOp` now implements `Copy`. Regarding the CI failures: You can wait for #5751 to get merged and then rebase-squash this PR on top of the master branch.

rail-rain commented:
Thank you for the information. I've subscribed #5751, so you don't have to ping me again when it gets merged. I will reflect the change in `AccocOp` after it because it's seemingly impossible to run the test with rust-lang/rust#73629 now for me.

bors commented:
:umbrella: The latest upstream changes (presumably #5751) made this pull request unmergeable. Please resolve the merge conflicts.

rail-rain commented:
@flip1995  I've run `git rebase master` and force-pushed it (although I was not sure if this is what you mean by saying 'rebase-squash'; should I squash my commits?). Also I've removed the manual `Clone` implementation on `Sugg` in favour of deriving. I think now it's finally ready for a review.

flip1995 commented:
> should I squash my commits?

Yes, I think that would be reasonable in that case, especially, since there are multiple `fmt` commits and small refactor commits. (Good for initial implementation+review, but doesn't add value to the history of the master branch).

bors commented:
:umbrella: The latest upstream changes (presumably #5763) made this pull request unmergeable. Please resolve the merge conflicts.

rail-rain commented:
Ah, I see. That perfectly make sense. However, I'm afraid squashing could really affects reviews (e.g. the `fmt` commits changed the level of indentation and hide what actually changed); has the review already done, or are you fine with reviewing one commit? And also, should I squash the changes I made after submitting the PR? It would break the references to the code in the past reviews (although I'm not sure if there would be anyone who would want to see these changes).

flip1995 commented:
Yeah, let's squash right before merging, this is a rather complex PR. I try to leave an in depth review in the coming days (maybe even today).

flip1995 commented:
I still didn't get to review this in-depth, and I'm sorry! I try to get to this soon.

rail-rain commented:
That's all right. Take your time. Actually, I don't feel so energetic recently and even haven't been working on other issues.

rail-rain commented:
Sorry, I didn't aware of the number of the changes made in `rustc` and Clippy. This rebase might confuse you.

flip1995 commented:
No problem, thanks for low key reminding me, that I still have to review this 😄

bors commented:
:umbrella: The latest upstream changes (presumably #6076) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

flip1995 reviewed (changes requested):
Sorry for taking so long to review this. Expect multiple review cycles for this lint.

> Secondly, manual copying with loop counters are likely to trigger `needless_range_loop` and `explicit_counter_loop` along with `manual_memcpy`; in fact, I explicitly allowed them in the tests. Is there any way to disable these two lints when a code triggers `manual_memcpy`?

Generally triggering two lints with conflicting suggestions at the same time is bad. Normally we can handle this, if they are in the same lint pass with something like this:

```
if !lint_a_linted {
    lint_b()
}
```

And then only run the lint B, if lint A didn't trigger. 



> And, another thing I'd like to note is that `Sugg` adds unnecessary parentheses when expressions with parentheses passed to its `hir` function, as seen here:

Don't worry about this for now. This is a easy fix in a second pass, because the `unnecessary_parens` lint will trigger, which is also auto-applicable.

---

The suggestion you currently generate only replaces the expression, that creates the range in the loop. You want to replace the whole loop. That means you need to use the span for the whole loop in the suggestion. You can keep the span you currently using for the lint message though (or make it so the span for the lint message starts with `for`).

rail-rain commented:
Hello,
Thanks a lot for reviwing!

I've addressed `snippet_opt` and conflicting lints.

Regarding the span for the suggestion, I've replaced the span with the whole loop; but I'm not sure if I've got what you meant by saying "You can keep the span you currently using". Can I use different spans for tools (namely `rustfix`) and messages? Having said that, I think the suggestion is fine as it is because other lints like `while_let_loop` uses this. (By the way, this weirdness have already existed for a while... [and `explicit_counter_loop` is also suffered from this](https://github.com/rust-lang/rust-clippy/pull/4691#issuecomment-543507847). [The original was good](https://github.com/rust-lang/rust-clippy/pull/2021/commits/90f345df94ffa92760ce338208a034d0a3785a2d), so I suspect rustup caused it at some point).

Oh, I see the test is falling because it's too long. It seems I need either to split up the test or to shorten the suggestion.

rail-rain commented on clippy_lints/src/loops.rs:
```diff
@@ -800,29 +800,126 @@ enum OffsetSign {
-struct FixedOffsetVar<'hir> {
-    var: &'hir Expr<'hir>,
-    offset: Offset,
+impl std::ops::Add<&MinifyingSugg<'static>> for MinifyingSugg<'static> {
+    type Output = MinifyingSugg<'static>;
+    fn add(self, rhs: &MinifyingSugg<'static>) -> MinifyingSugg<'static> {
```
Um... I guss it's because not being generic was a lot easier and it's enough for this case - the content of underlying `Cow` is either `&'static str` (default snippets) or `String` (mainly from `format!`). `MinifyingSugg` sometimes returns `self`; as opposed to `Sugg`, which always create a `String` and return `Sugg<'static>` regardless the arguments' lifetime. The correct way would probably be defining lifetimes like `<'lhs: 'output, 'rhs: 'output, 'output>`

rail-rain commented on clippy_lints/src/loops.rs:
```diff
@@ -800,29 +800,126 @@ enum OffsetSign {
+}
+
+#[derive(Clone)]
+struct MinifyingSugg<'a>(sugg::Sugg<'a>);
```
I'm not sure if I agree. Firstly, for `self.0`, most of `self.0` (and `rhs.0`) are where `self` is a reference so it would become `&**self`; and, two `rhs.0.clone()` would be `(**rhs).clone()` due to their precedence. Secondly, there're just few places where auto-dereferencing could happen; probably, only when I made `MinifyingSugg::as_str` the method of `Sugg`.

flip1995 replied:
Alright, I'm convinced 👍

flip1995 commented:
Changes LGTM, I will do a complete review again right after writing this comment. I just want to write down my thoughts before forgetting them.

> Can I use different spans for tools (namely rustfix) and messages?

Yes, in Clippy with `span_lint_and_then(LINT, span1, ..., |diag| { diag.span_suggestion(span2, sugg) })`. In the `|diag|` closure you can then use the full potential of the [`DiagnosticBuilder`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.DiagnosticBuilder.html). 

But seeing the suggestion as it is now, I don't think this is necessary and the suggestion is fine now.

> I've split the test for now; I might revert it if the suggestion became much shorter later.

Splitting the tests definitely make sense here. I would keep it that way.

flip1995 commented on clippy_lints/src/loops.rs:
```diff
@@ -845,151 +946,207 @@ fn fetch_cloned_expr<'tcx>(expr: &'tcx Expr<'tcx>) -> &'tcx Expr<'tcx> {
-                match offset.sign {
-                    OffsetSign::Negative => format!("({} - {})", snippet(cx, end.span, "<src>.len()"), offset.value),
-                    OffsetSign::Positive => "".into(),
+    let print_limit =
+        |end: &Expr<'_>, end_str: &str, base: &Expr<'_>, sugg: MinifyingSugg<'static>| -> MinifyingSugg<'static> {
```
Is there a reason, that this is a closure and not a function? Scanning over the code I couldn't find any upvars, so it should be possible to make it a function for consistency.

rail-rain replied:
`print_limit` captures `cx` and reduces the number of its argument. I actually prefer to make it a function; I kept it just because it was, and there's a couple of examples of this practice in Clippy.

rail-rain replied:
`print_limit` captures `cx` and reduces the number of its argument. I actually prefer to make it a function; I kept it just because it was, and there's a couple of examples of this practice in Clippy.

flip1995 replied:
Ah ok. I'm good with both, so you decide if you want to turn it into a function 👍

rail-rain replied:
It turned out `print_limit` also captures `limits`, and `print_offset_and_limit` captures so many; and then I changed my mind. I'd like to keep these closures. (I even removed the explicit return type from `print_limit`)

flip1995 reviewed (commented):
Round 2/3?: Mostly documentation and some style things I noticed.

rail-rain commented on clippy_lints/src/loops.rs:
```diff
@@ -845,151 +946,207 @@ fn fetch_cloned_expr<'tcx>(expr: &'tcx Expr<'tcx>) -> &'tcx Expr<'tcx> {
+fn get_assignments<'a: 'c, 'tcx: 'c, 'c>(
+    cx: &'a LateContext<'tcx>,
+    stmts: &'tcx [Stmt<'tcx>],
+    expr: Option<&'tcx Expr<'tcx>>,
```
I have no idea 😳. It's probably an artifact from refactoring before submitting.

flip1995 commented on tests/ui/manual_memcpy/with_loop_counters.rs:
```diff
@@ -0,0 +1,71 @@
+    for i in 3..src.len() {
+        dst[i] = src[count];
+        count += 1
+    }
```
This should be the last review round:

I would like two more tests (correct me if they won't make sense):

1. Use `dst.len()` instead of `src.len()`
2. put `count += 1;` in front of `dst[i] = src[count];`

rail-rain replied:
I probably thought it was fine because the `dst.len()` case is tested in `without_loop_counters.rs`, and the 'loop counters at the top of the loop' (am I understand what you said correctly?) case is the almost same as #6076, a FP in `expilcit_counter_loop` which shares the same visitor to check this case. Having said that, I cannot say these cases don't make sense; so I will add these.

rail-rain commented on clippy_lints/src/loops.rs:
```diff
@@ -845,151 +946,207 @@ fn fetch_cloned_expr<'tcx>(expr: &'tcx Expr<'tcx>) -> &'tcx Expr<'tcx> {
-            ExprKind::Path(..) if !same_var(cx, e, var) => Some(snippet_opt(cx, e.span).unwrap_or_else(|| "??".into())),
+            ExprKind::Path(..) if get_start(cx, e, starts).is_none() => {
+                // `e` is always non paren as it's a `Path`
+                Some(Sugg::NonParen(snippet(cx, e.span, "???")))
```
I've avoided using it because I thought it was unnecessary, but I don't have a strong opinion here

flip1995 commented:
Thanks for all the great work! And for addressing my reviews so quickly, even though I was really slow myself.

@bors r+ Let's finally merge this 🚀

bors commented:
:pushpin: Commit b54188429409a6da5f931fa4be5df1f40b377015 has been approved by `flip1995`

<!-- @bors r=flip1995 b54188429409a6da5f931fa4be5df1f40b377015 -->
<!-- homu: {"type":"Approved","sha":"b54188429409a6da5f931fa4be5df1f40b377015","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit b54188429409a6da5f931fa4be5df1f40b377015 with merge dbc02854fce1f783599e3f98b6ee6b9a136d9e96...
<!-- homu: {"type":"BuildStarted","head_sha":"b54188429409a6da5f931fa4be5df1f40b377015","merge_sha":"dbc02854fce1f783599e3f98b6ee6b9a136d9e96"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/1234710048), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/1234708883), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1234743443)
Approved by: flip1995
Pushing dbc02854fce1f783599e3f98b6ee6b9a136d9e96 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/1234743443","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/1234710048","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/1234708883"},"merge_sha":"dbc02854fce1f783599e3f98b6ee6b9a136d9e96"} -->
